### DISCUSSION TITLE
Exporting a kernel into QPU compatible OpenQASM2 file

### DISCUSSION THREAD

**Discussion by ashsaki** [NVIDIA/cuda-quantum#2005] [Q&A] [answered]
Suppose I have a kernel with high-level gates (that may not adhere to target QPU's basis gates and qubit connectivity). I want to get a QPU-compatible version of the kernel in OpenQASM2 format. 

For example, I was trying to compile the following code for IQM Adonis QPU.
```
#include <cudaq.h>
#include <fstream>

struct adonis_kernel {
  // Adonis QPU.
  //       QB1
  //        |
  // QB2 - QB3 - QB4
  //        |
  //       QB5

  void operator()() __qpu__ {
    cudaq::qvector q(5);
    x(q[0]);

    for (int i = 0; i < 4; i++) {
      x<cudaq::ctrl>(q[i], q[i + 1]);
    }
    x<cudaq::ctrl>(q[0], q[2], q[1]);
    auto result = mz(q);
  }
};

int main() {
  // auto counts = cudaq::sample(adonis_kernel{});
  // counts.dump();
}
```

The `nvq++` compiler flags I used:
```
nvq++ --emulate --target iqm --iqm-machine Adonis iqm_simple.cpp -o iqm_simple.x -v --save-temps
```
The `-v` shows that the `.cpp` code is converted to a `quake (.qke)` IR. Then, `cudaq-opt` runs on the `.qke` file, and the `cudaq-opt` use a pipeline of passes. Then, `cudaq-translate` is applied. 

I tried to manually run the `cudaq-translate` with `--convert-to=qasm2`, But, it threw me an error. 

From browsing the source code, one can see pipelines of optimization passes and tools for this conversion.

- How can one use the capabilities available in `cudaq` to generate QPU compatible OpenQASM representation of the a kernel?

    -- While I used IQM as an example for trial purpose, it will be helpful if the solution can can take user-defined basis gates and qubit connectivity.
   -- Sample or Observe the kernel is not a necessity now for me.

**Comment by annagrin** [comment, marked as answer]
Thank you [ashsaki](https://github.com/ashsaki) for the example, that discovered an issue with our current pipeline (some missing opts in conversion to OpenQASM2).

Here is a workaround that works for this concrete example, please let us know if you have any other failing ones:

```
cudaq-quake iqm_simple.cpp | cudaq-opt --pass-pipeline="builtin.module(func.func(memtoreg{quantum=0}),cc-loop-unroll)" | cudaq-translate --convert-to=openqasm2
```

**Comment by annagrin** [comment]
Filed an issue: https://github.com/NVIDIA/cuda-quantum/issues/2176

**Comment by ashsaki** [comment]
Thanks!

To add to your answer, I found the following is needed to address basis conversion and routing on a restricted qubit connectivity.

```
cudaq-opt -pass-pipeline='builtin.module(func.func(multicontrol-decomposition),basis-conversion{basis=h,rz,x,x(1)},func.func(add-dealloc,combine-quantum-alloc,canonicalize,factor-quantum-alloc,memtoreg,qubit-mapping{device=path(10)},regtomem),basis-conversion{basis=h,rz,x,x(1)})'  out.qke > optimized.qke
```

- The qubit connectivity can be user-defined following a certain format (e.g., [IQM Adonis](https://github.com/NVIDIA/cuda-quantum/blob/3dd2efd5b975ecf35349a21733d5f168f711382a/runtime/cudaq/platform/default/rest/helpers/iqm/Adonis.txt)). 
- I found that *basis* cannot take any arbitrary universal set yet.
